### ISSUE TITLE
C++ Library Working

### ISSUE THREAD

**Issue by r-barnes** [gerrymandr/python-mander#11]
The `cpp` branch, which should be available on Github now, is integrated into Python.

The C++ code is currently in a repo at "https://github.com/r-barnes/compactnesslib".

I haven't yet set the library to compile using OpenMP (which would parallelize everything), but could do so in the future, though it may complicate the build somewhat.

The cpp branch does not yet replace any of the existing functionality of the library, but I'm raising this issue to suggest that we modify the library so that it does.

An example of how to use compactnesslib from Python is as follows:

    #!/usr/bin/env python

    import mander
    import sys

    if len(sys.argv)!=2:
      print("Syntax: {0} <GeoJSON file>".format(sys.argv[0]))
      sys.exit(-1)

    with open(sys.argv[1], 'r') as fin:
      content    = fin.read()
      score_json = mander.ProcessGeoJSON(content, '')
      print(score_json)

Handing this an input GeoJSON consisting of Massachusetts's electoral districts returns, after 0.07s on my laptop, the following output:

    {
      "0":{
        "perim":500138,
        "area":6.21574e+09,
        "PolsbyPopp":0.312265,
        "Schwartzbe":0.558807,
        "ConvexHull":0.74883,
        "Reock":0.410659
      },
      "1":{
        "perim":452510,
        "area":4.4387e+09,
        "PolsbyPopp":0.272402,
        "Schwartzbe":0.521921,
        "ConvexHull":0.770146,
        "Reock":0.398661
      },
      "2":{
        "perim":335527,
        "area":2.03904e+09,
        "PolsbyPopp":0.227605,
        "Schwartzbe":0.477079,
        "ConvexHull":0.675721,
        "Reock":0.31953
      },
      "3":{
        "perim":463187,
        "area":1.79412e+09,
        "PolsbyPopp":0.105087,
        "Schwartzbe":0.324171,
        "ConvexHull":0.64404,
        "Reock":0.366629
      },
      "4":{
        "perim":204721,
        "area":1.67266e+08,
        "PolsbyPopp":0.0501522,
        "Schwartzbe":0.223947,
        "ConvexHull":0.567434,
        "Reock":0.072261
      },
      "5":{
        "perim":498935,
        "area":8.67639e+08,
        "PolsbyPopp":0.0437988,
        "Schwartzbe":0.209282,
        "ConvexHull":0.567975,
        "Reock":0.191279
      },
      "6":{
        "perim":2.01532e+06,
        "area":3.3033e+09,
        "PolsbyPopp":0.0102205,
        "Schwartzbe":0.101096,
        "ConvexHull":0.551033,
        "Reock":0.0405621
      },
      "7":{
        "perim":662617,
        "area":1.41659e+09,
        "PolsbyPopp":0.0405442,
        "Schwartzbe":0.201356,
        "ConvexHull":0.713262,
        "Reock":0.174989
      },
      "8":{
        "perim":252726,
        "area":7.1762e+08,
        "PolsbyPopp":0.141191,
        "Schwartzbe":0.375753,
        "ConvexHull":0.613872,
        "Reock":0.294034
      }
    }

The output ids could be selected from the incoming data, though I'm not sure what the attribute names are or how they are stored in the objects. Could someone upload an example GeoJSON?

The output could also be the fully GeoJSON multipolygons with the scores attached as attributes.

If I don't hear back from anyone in a week or so, I'll likely make an attempt at modifying the Python myself.

Hope everyone's doing well!

**Comment by mxndrwgrdnr** [comment]
Hey @r-barnes, there are some GeoJSON files in the test/data directory of the repo. As for the structure of the output of `metrics.scoresToGeojson()`, the specified metric name and score gets added as a key-value pair to the `properties` attribute of the original .geojson file. Let me know if you need any more detail than that.

**Comment by r-barnes** [comment]
Thanks, I've got property extraction from GeoJSON working now.

I've also, as of today, got bounded scores working. These take in the boundaries of electoral districts and the boundaries of the states which contain them and calculate compactness scores which account for the fact that the electoral district must be contained by the state.

Since the interface uses GeoJSON or WKT to communicate with the library I've also implemented a caching scheme to reduce computation.

It's probably worth chatting about what good return values are, and what the interface should look like.

At the moment the library ingests GeoJSON and returns a JSON dictionary containing only the scores for each district queried. I could imagine it ingesting shapely items and converting them to library format internally, as well as returning JSON dictionaries, attributified GeoJSON, or straight-out Python dictionaries.

The library also includes some high-performance, bulk-processing functionality to read/write shapefiles directly from disk without involving Python. This allows for commands like:

    mander.addBoundedScoresToNewShapefile(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])

Use cases are needed at this point.
